python/pyarrow/dataset.py

```diff
@@ -232,168 +219,405 @@ def _ensure_format(obj):
         raise ValueError("format '{}' is not supported".format(obj))
 
 
-def factory(path_or_paths, filesystem=None, partitioning=None,
-            format=None):
+def _ensure_filesystem(fs_or_uri):
```

Comment:
Can be for a follow-up JIRA, but we might want to move this helper to `pyarrow.fs`, and use some this also in other places where we accept filesystems? (although I actually don't know if there are already many such places ..)